// convex/dist/esm/cli/lib/localDeployment/filePaths.js
"use strict";
import path from "path";
import { rootDirectory } from "../utils/utils.js";
import { logVerbose } from "../../../bundler/context.js";
import { recursivelyDelete } from "../fsUtils.js";
export function rootDeploymentStateDir(kind) {
  return path.join(
    rootDirectory(),
    kind === "local" ? "convex-backend-state" : "tryitout-convex-backend-state"
  );
}
export function deploymentStateDir(deploymentKind, deploymentName) {
  return path.join(rootDeploymentStateDir(deploymentKind), deploymentName);
}
export function loadDeploymentConfig(ctx, deploymentKind, deploymentName) {
  const dir = deploymentStateDir(deploymentKind, deploymentName);
  const configFile = path.join(dir, "config.json");
  if (!ctx.fs.stat(dir).isDirectory()) {
    return null;
  }
  if (ctx.fs.exists(configFile)) {
    const content = ctx.fs.readUtf8File(configFile);
    try {
      return JSON.parse(content);
    } catch (e) {
      logVerbose(ctx, `Failed to parse local deployment config: ${e}`);
      return null;
    }
  }
  return null;
}
export function saveDeploymentConfig(ctx, deploymentKind, deploymentName, config) {
  const dir = deploymentStateDir(deploymentKind, deploymentName);
  const configFile = path.join(dir, "config.json");
  if (!ctx.fs.exists(dir)) {
    ctx.fs.mkdir(dir, { recursive: true });
  }
  ctx.fs.writeUtf8File(configFile, JSON.stringify(config));
}
export function binariesDir() {
  return path.join(rootDirectory(), "binaries");
}
export function binaryZip() {
  return path.join(binariesDir(), "convex-backend.zip");
}
export function dashboardZip() {
  return path.join(dashboardDir(), "dashboard.zip");
}
export function versionedBinaryDir(version) {
  return path.join(binariesDir(), version);
}
export function executablePath(version) {
  const ext = process.platform === "win32" ? ".exe" : "";
  return path.join(versionedBinaryDir(version), `convex-local-backend${ext}`);
}
export function dashboardDir() {
  return path.join(rootDirectory(), "dashboard");
}
export async function resetDashboardDir(ctx) {
  const dir = dashboardDir();
  if (ctx.fs.exists(dir)) {
    await recursivelyDelete(ctx, dir);
  }
  ctx.fs.mkdir(dir, { recursive: true });
}
export function dashboardOutDir() {
  return path.join(dashboardDir(), "out");
}
export function loadDashboardConfig(ctx) {
  const configFile = path.join(dashboardDir(), "config.json");
  if (!ctx.fs.exists(configFile)) {
    return null;
  }
  const content = ctx.fs.readUtf8File(configFile);
  try {
    return JSON.parse(content);
  } catch (e) {
    logVerbose(ctx, `Failed to parse dashboard config: ${e}`);
    return null;
  }
}
export function saveDashboardConfig(ctx, config) {
  const configFile = path.join(dashboardDir(), "config.json");
  if (!ctx.fs.exists(dashboardDir())) {
    ctx.fs.mkdir(dashboardDir(), { recursive: true });
  }
  ctx.fs.writeUtf8File(configFile, JSON.stringify(config));
}
